exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,970 @@
+# Function Composition
+
+Status: **Proposed**
+
+<details>
+	<summary>Metadata</summary>
+	<dl>
+		<dt>Contributors</dt>
+		<dd>@catamorphism (also see Acknowledgments)</dd>
+		<dt>First proposed</dt>
+		<dd>2024-03-26</dd>
+		<dt>Pull Requests</dt>
+		<dd>#753</dd>
+	</dl>
+</details>
+
+## Objective
+
+_What is this proposal trying to achieve?_
+
+### Non-goal
+
+The objective of this design document is not to make
+a concrete proposal, but rather to explore a problem space.
+This space is complicated enough that agreement on vocabulary
+is desired before defining a solution.
+
+Instead of objectives, we present a primary problem
+and a set of subsidiary problems.
+
+### Problem statement: defining resolved values
+
+The problem defined in this design document is that
+the meaning of function composition is ambiguous.
+
+An obstacle to disambiguating its meaning
+is the absence of a definition of "resolved value".
+
+So we must address both problems together:
+
+* Problem 1: Define what it means for functions to compose with each other.
```

Comment:
> I don't know that this is actually a problem. I think it is a side-effect of not resolving Problem 2.

I disagree with this, because I think that even if we resolved Problem 2, we would still have ambiguities.

In your first example: suppose we solve problem 2 by adopting a representation of function results that includes all the options that were passed in. Without further spec work, though, it's _still_ unclear whether `$t` is printed with the time zone or not. Problem 2 is "what do functions return?", while problem 1 is "what do functions do with their inputs?" Maybe the latter question can only be answered in a way that's specific to each function.

Reply:
I don't think example 1 is an exemplar of "what do functions do with their inputs" as I don't think that options are necessarily the same thing as "inputs". Also, it's probably not a great example, since Temporal types provide ways to float/unfloat time values (making the time zone a part of the datetime value--that it, modifying the operand, as it were) rather than merely being a random option.

My concern is that, whenever I see "we have to solve two problems at the same time", that suggests extra effort to ensure that there's been complete decomposition of the requirements.

Reply:
Re: decomposition of the requirements, fair enough... what do you think about the change to this section I made in [86c26ff](https://github.com/unicode-org/message-format-wg/pull/753/commits/86c26ff8343913a2d46a1ab5b2cf31c98709a2b1) (emphasizing that this document is mainly about solving the resolved value problem, which is only a part of the bigger function composition problem)?